File: Day_3/backend/api/contractController.js
const { where } = require("sequelize");
const Contract = require("../models/Contract");
const { sendFCM } = require("../service_firebase/fcm_sender");

class ContractController {
  // Lấy tất cả hợp đồng
  static async getAllContracts(req, res) {
    try {
      const contracts = await Contract.findAll();
      res.status(200).json(contracts);
      console.log("lấy danh sách toàn bộ hợp đồng thành công", new Date());
    } catch (error) {
      res.status(500).json({ message: "lỗi lấy danh sách", error });
    }
  }

  // Lấy hợp đồng theo user_id
  static async getContractsByUser_Id(req, res) {
    try {
      const contracts = await Contract.findAll({
        where: { user_id: req.params.user_id },
      });
      res.status(200).json(contracts);
      console.log("lấy thành công", new Date());
    } catch (error) {
      console.error("lỗi khi lấy ds:", error);
      res.status(500).json({ message: "lỗi khi lấy ds:", error });
    }
  }

  // Thêm hợp đồng mới
  static async createContract(req, res) {
    try {
      const { user_id, contract_type, start_date, end_date, status } = req.body;
      const created_at = new Date();

      const newContract = await Contract.create({
        user_id,
        contract_type,
        start_date,
        end_date,
        status,
        created_at,
      });

      res.status(201).json({
        message: "tạo thành công",
        contract: newContract,
      });
    } catch (error) {
      res.status(500).json({ message: "Lỗi khi tạo: ", error });
    }
  }

  // Xoá hợp đồng
  static async deleteContract(req, res) {
    try {
      const { contract_id } = req.params;
      const deleted = await Contract.destroy({ where: { contract_id } });

      if (deleted) {
        res.json({ message: "Xóa thành công" });
      } else {
        res.status(404).json({ message: "hợp đồng không tồn tại" });
      }
    } catch (error) {
      res.status(500).json({ message: "xóa thất bại", error });
    }
  }

  // Sửa hợp đồng
  static async updateContract(req, res) {
    try {
      const { contract_id } = req.params;
      const { user_id, contract_type, start_date, end_date, status } = req.body;

      const [updated] = await Contract.update(
        { user_id, contract_type, start_date, end_date, status },
        { where: { contract_id } }
      );

      if (updated) {
        res.json({ message: "Đã cập nhật thành công" });
      } else {
        res.status(404).json({ message: "không tồn tại hợp đồng" });
      }
    } catch (error) {
      res.status(500).json({ message: "cập nhật lỗi", error });
    }
  }

  static async getContractByID(req, res) {
    try {
      const { contract_id } = req.params;

      const contract = await Contract.findOne({
        where: { contract_id },
      });

      if (contract) {
        res.status(200).json(contract);
        console.log(`Lấy thành công hợp đồng ID: ${contract_id}`, new Date());
      } else {
        res.status(404).json({ message: "Không tìm thấy hợp đồng" });
      }
    } catch (error) {
      console.error(
        `Lỗi khi lấy hợp đồng ID: ${req.params.contract_id}`,
        error
      );
      res.status(500).json({
        message: "Lỗi khi lấy thông tin hợp đồng",
        error: error.message,
      });
    }
  }
}

module.exports = ContractController;

File: Day_3/my_app/lib/database/contract_controller.dart
// db/contract_controller.dart
import 'package:isar/isar.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/contract.dart';
import 'db_helper.dart';

class ContractController {
  /// Thêm một hợp đồng
  static Future<void> addContract(Contract contract,
      {String status = "created"}) async {
    final prefs = await SharedPreferences.getInstance();
    final userId = int.tryParse(prefs.getString('user_id') ?? "0") ?? 0;

    Contract newContract = Contract(
      userId: userId,
      contractType: contract.contractType,
      startDate: contract.startDate,
      endDate: contract.endDate,
      status: contract.status,
      createdAt: contract.createdAt,
      syncStatus: status,
    );

    if (contract.contractId != 0) newContract.contractId = contract.contractId;

    final isarInstance = await DBHelper.isar;
    await isarInstance.writeTxn(() async {
      await isarInstance.contracts.put(newContract);
    });
  }

  /// Thêm nhiều hợp đồng
  static Future<void> saveContracts(List<Contract> contracts) async {
    final isarInstance = await DBHelper.isar;
    await isarInstance.writeTxn(() async {
      await isarInstance.contracts.putAll(contracts);
    });
    print(
        "Đã lưu ${await isarInstance.contracts.count()} hợp đồng vào database.");
  }

  /// Lấy tất cả hợp đồng
  static Future<List<Contract>> getAllContracts() async {
    final isarInstance = await DBHelper.isar;
    final allContracts = await isarInstance.contracts.where().findAll();
    return allContracts.where((c) => c.syncStatus != 'deleted').toList();
  }

  /// Lấy danh sách hợp đồng theo userId
  static Future<List<Contract>> getContractsByUserId(int userId) async {
    final isarInstance = await DBHelper.isar;
    return await isarInstance.contracts
        .filter()
        .userIdEqualTo(userId)
        .findAll();
  }

  static Future<void> updateContract(Contract updatedContract,
      {String status = "updated"}) async {
    updatedContract.syncStatus = status;

    final isarInstance = await DBHelper.isar;
    await isarInstance.writeTxn(() async {
      await isarInstance.contracts.put(updatedContract);
    });
  }

  /// Xóa hợp đồng theo ID
  static Future<void> deleteContract(int id,
      {String status = "deleted"}) async {
    final isarInstance = await DBHelper.isar;
    final contract = await isarInstance.contracts.get(id);
    if (contract != null) {
      if (status == "deleted") {
        contract.syncStatus = status;
        await isarInstance.writeTxn(() async {
          await isarInstance.contracts.put(contract);
        });
      } else {
        await isarInstance.writeTxn(() async {
          await isarInstance.contracts.delete(id);
        });
      }
    }
  }

  //clear
  static Future<void> clearContract() async {
    final isarInstance = await DBHelper.isar;

    await isarInstance.writeTxn(() async {
      await isarInstance.contracts.clear();
    });
  }

  //lấy các hợp đồng cần đồng bô
  static Future<List<Contract>> getUnsyncedContracts() async {
    final isarInstance = await DBHelper.isar;

    // Lấy tất cả hợp đồng từ local DB
    final allContracts = await isarInstance.contracts.where().findAll();

    // Lọc thủ công bằng Dart
    return allContracts.where((contract) {
      return contract.syncStatus != 'synced';
    }).toList();
  }
}
